tools/vhost-user-video/v4l2_backend.c

```diff
@@ -1013,51 +1034,80 @@ void v4l2_print_event(const struct v4l2_event *ev)
     }
 }
 
-int v4l2_queue_buffer(int fd, enum v4l2_buf_type type,
+int v4l2_queue_buffer(enum v4l2_buf_type type,
+                      enum v4l2_memory memory,
                       struct virtio_video_resource_queue *qcmd,
                       struct resource *res, struct stream *s,
-                      struct v4l2_device *dev)
+                      struct v4l2_device *dev, struct vuvbm_device *bm_dev)
 {
     struct v4l2_buffer vbuf;
     int ret = 0;
+    int fd = s->fd;
 
     memset(&vbuf, 0, sizeof(vbuf));
     vbuf.index = res->v4l2_index;
 
     vbuf.type = type;
+    vbuf.memory = memory;
     vbuf.field = V4L2_FIELD_NONE;
     vbuf.flags = V4L2_BUF_FLAG_TIMESTAMP_COPY;
 
-    g_debug("%s: type=%s index=%d", __func__,
-            v4l2_buf_type_name(type), vbuf.index);
+    g_debug("%s: type=%s memory=%d index=%d", __func__,
+            v4l2_buf_type_name(type), memory, vbuf.index);
 
     convert_to_timeval(le64toh(qcmd->timestamp), &vbuf.timestamp);
 
-    /* if using GUEST_PAGES queued using USERPTR mechanism */
-    vbuf.memory = V4L2_MEMORY_USERPTR;
-
-    if (video_is_mplane(type)) {
-        /* for mplane length field is number of elements in planes array */
-        vbuf.length = res->vio_resource.num_planes;
-        vbuf.m.planes = g_malloc0(sizeof(struct v4l2_plane)
-                                  * res->vio_resource.num_planes);
+    if (memory == V4L2_MEMORY_USERPTR) {
+        if (video_is_mplane(type)) {
+            /* for mplane length field is number of elements in planes array */
+            vbuf.length = res->vio_resource.num_planes;
+            vbuf.m.planes = g_malloc0(sizeof(struct v4l2_plane)
+                                    * res->vio_resource.num_planes);
 
-        for (int i = 0; i < vbuf.length; i++) {
-            vbuf.m.planes[i].m.userptr = (unsigned long)res->iov[i].iov_base;
-            vbuf.m.planes[i].length = (unsigned long)res->iov[i].iov_len;
+            for (int i = 0; i < vbuf.length; i++) {
+                vbuf.m.planes[i].m.userptr = (unsigned long)res->iov[i].iov_base;
+                vbuf.m.planes[i].length = (unsigned long)res->iov[i].iov_len;
+            }
+        } else if (res->iov != NULL) {
+            vbuf.m.userptr = (unsigned long)res->iov[0].iov_base;
+            vbuf.length = res->iov[0].iov_len;
+            g_debug("%s: iov_base = 0x%p", __func__, res->iov[0].iov_base);
+            g_debug("%s: iov_len = 0x%lx", __func__, res->iov[0].iov_len);
+        }
+    } else {
+        assert(memory == V4L2_MEMORY_DMABUF);
+        if (video_is_mplane(type)) {
+            g_printerr("MPLANE not supported with DMABUF memory.");
```

Comment:
Yes, it can be supported, it is just that the patch does not support it... yet.

I am focusing on single planes as it is what I can test in my laptop with `vicodec`. Once I have the setup for the Chromebook ready, I will add the support for `MPLANES`. The print is a reminder. Maybe I can add a TODO comment to remark it.